app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt

```diff
@@ -400,6 +401,21 @@ class QuestionPlayerFragmentPresenter @Inject constructor(
     )
   }
 
+  private val keyboardVisibilityListener = ViewTreeObserver.OnGlobalLayoutListener {
+    val diff = binding.root.rootView.height - binding.root.height
+    val threshold = 200 * activity.resources.displayMetrics.density // represents 200 dp
+    val isKeyboardVisible = diff > threshold
+    questionViewModel.isKeyboardVisible.set(isKeyboardVisible)
```

Comment:
This seems like a really hacky way to check if the keyboard is open. It's making a lot of assumptions on how large the keyboard is (which is a customizable and OEM-specific property). I think the discussion in https://groups.google.com/g/android-platform/c/FyjybyM0wGA is insightful--it's better practice to build a solution that doesn't directly depend on the keyboard itself. I think this is a case where Android is intentionally designed to help avoid developers inadvertently introducing OEM-specific bugs.

In this case, could we change how we display the progress bar such that it wraps under the keyboard, but stays on top of the recycler view when scrolling? That seems to me to be the more correct solution rather than trying to tweak the UI behavior on the keyboard itself.

@rt4914 in case you have any ideas on how to do the above.

Reply:
@BenHenning @rt4914 
> I think this is a case where Android is intentionally designed to help avoid developers inadvertently introducing OEM-specific bugs.

Actually, Android Team is working on solutions to provide developers with more control over IME options. In fact, Android 11 has a new API `WindowInsets` dedicated to this. The API provides a function to check if the keyboard is visible or not. (https://youtu.be/acC7SR1EXsI?list=PLWz5rJ2EKKc-Z8NeBXJkf1bzUVhx3fvh4&t=299). I couldn't use this because of supported SDK version constraints.

> In this case, could we change how we display the progress bar such that it wraps under the keyboard, but stays on top of the recycler view when scrolling? 

The keyboard doesn't come over the fragment, it pushes the fragment up. So, I doubt if we can do something like hiding/moving the progress bar "behind" the keyboard.

Reply:
@BenHenning I do agree with your point of view. 
@ArpitShukIa What can be the best solution for the issue #2131  without doing all this calculations?

Reply:
@BenHenning @rt4914
I am not sure what can be the best solution for this.
We don't have a lot of options while dealing with soft-keyboard. This solution has a lot of upvotes in [stackoverflow](https://stackoverflow.com/questions/2150078/how-to-check-visibility-of-software-keyboard-in-android).

One modification we can do is something like 
`val isKeyboardVisible = binding.root.rootView.height > 1.5 * binding.root.height` 
But yeah, here also we will be assuming that if more than 1/3rd of the screen is covered, it might be due to soft keyboard.
I don't think there is any straightforward way to check keyboard's visibility (before API 30)

I also tried using `android:windowSoftInputMode="adjustPan"` for the activity so that the keyboard comes above the activity and doesn't push it up. The progress bar was hiding correctly behind the keyboard but there was a problem that some part of ScrollView was also hiding behind it and couldn't be scrolled to view the bottom contents.

I can't think of any other way to approach the problem. Please suggest if you have any ideas.

Reply:
Sorry to be clear: I'm suggesting we try to solve this problem a different way that does not require knowing whether the keyboard is open. Could we explore changing the UI itself to address the issue first? It may require a part UI redesign depending on what options we have, but that might be fine. I'm worried about the stability of this solution on different devices--it will be quite difficult for us to know if something goes wrong since the user won't necessary understand that it's broken to report it to us. Given that, we should try to find a way to solve this through the UI structure itself first before trying to factor in the size of the keyboard.

Reply:
I can try modifying the UI (something like positioning the progress bar on top) but I think we should take some inputs from the UI Designer first regarding any UI change.

Reply:
@mschanteltc Can you suggest some UI changes to fix this issue #2131?

Reply:
The progress bar should always be fixed to the bottom, so when the keyboard is active, it should cover the progress bar.

Reply:
@BenHenning @rt4914 Need some help here. Can you tell exactly what UI changes need to be done?

Reply:
@mschanteltc we might need to change the UI experience in another way to make this work. We'll follow up if that's the case.

@rt4914 we should discuss options further here. I'm wondering if there's some way to tweak the UI such that the progress bar stays above other elements but doesn't get pushed up by the keyboard.

Reply:
@BenHenning Ack. But I am actually not sure what would be the best solution here. I will try to use this PR to see if I can do something.

Reply:
From talking to @rt4914 today, we were discussing of two options:
- Experimenting with switching this layout to adjustPan instead of adjustResize. This might help circumvent the issue, but will worsen the user experience.
- Adding [configuration handling support](https://developer.android.com/guide/topics/resources/runtime-changes#HandlingTheChange) (e.g. via ``keyboardHidden``) to this activity for keyboard changes & then look at the Configuration class for [whether the software keyboard is open](https://developer.android.com/reference/android/content/res/Configuration#keyboardHidden) (& condition the visibility of the progress bar on that)

@ArpitShukIa it might be worth trying (2) and see if that's a viable option here.

Reply:
@BenHenning @rt4914 
(2) doesn't seem to be working. I added `android:configChanges="keyboardHidden"` in the manifest for QuestionPlayerActivity and overrode onConfigurationChanged() in the activity. But that function is not being triggered when the visibility of softKeyboard changes.

I think a soft keyboard doesn't trigger activity restart on its appearance or disappearance. Looks like the flag just tracks the "availability" of soft and hard keyboard and not the "visibility" (in case of soft keyboard). (The name `keyboardHidden` seems to be a little misleading)

![image](https://user-images.githubusercontent.com/51705072/115949270-89bef980-a4f1-11eb-8d75-07de63c10b7b.png)

Did I miss something here?

Reply:
Hmm no I think you're right. How does the experience look when the activity is switched to adjustPan mode @ArpitShukIa?

We also might be able to use https://developer.android.com/jetpack/androidx/releases/core#1.5.0-alpha02. There's a new API that lets us check the visibility of the onscreen keyboard in hopefully a more robust way than directly checking its size (since per the stackoverflow page is easy to mess up). However, that API will also require us to use adjustPan so we should check that experience first.

If possible, could you update this PR with the existing solution commented out & adjustPan enabled? The CI checks automatically build an APK that we can download locally & check quickly to see the experience.

Reply:
@BenHenning I have updated the PR to use adjustPan. PTAL.
The behaviour is the same as I described in [this](https://github.com/oppia/oppia-android/pull/2933#discussion_r606179734) comment.

https://user-images.githubusercontent.com/51705072/116185662-ae101580-a73f-11eb-99f3-df3eab2c556a.mp4

Reply:
@rt4914 & @mschanteltc do either of you have thoughts on the most recent video provided by @ArpitShukIa? Either we need to redesign the question player to work better with the keyboard, or we might need to live with adjustPan since it's a limitation of Android.

Also, in case it wasn't clear in the video, in some cases the user will not be able to access the submit button because of the keyboard being on top of it. We *might* be able to make this a bit better by detecting using the library above when the keyboard is open and dynamically add more space to offset it (@rt4914 & @ArpitShukIa any thoughts on this?).

Reply:
@rt4914 & @mschanteltc do either of you have thoughts on the most recent video provided by @ArpitShukIa? Either we need to redesign the question player to work better with the keyboard, or we might need to live with adjustPan since it's a limitation of Android.

Also, in case it wasn't clear in the video, in some cases the user will not be able to access the submit button because of the keyboard being on top of it. We *might* be able to make this a bit better by detecting using the library above when the keyboard is open and dynamically add more space to offset it (@rt4914 & @ArpitShukIa any thoughts on this?).

Reply:
> We might be able to make this a bit better by detecting using the library above when the keyboard is open

I tried using the API but am unable to find any function which can be used for API 19 and 20 (as in our case).
The two ways that I tried are:
1. Using `ViewCompat.setOnApplyWindowInsetsListener(View, OnApplyWindowInsetsListener)`
```kotlin
ViewCompat.setOnApplyWindowInsetsListener(binding.root.rootView) { _, insets ->
      val isKeyboardVisible = insets.isVisible(Type.ime())
      val keyboardHeight = insets.getInsets(Type.ime()).bottom
      insets
}
```
But this function works only for API>=21
![image](https://user-images.githubusercontent.com/51705072/116390754-00366100-a83c-11eb-9832-953f9ae777f5.png)
<br>
2. Using `ViewCompat.getRootWindowInsets(View)` inside `ViewTreeObserver.OnGlobalLayoutListener.onGlobalLayout()` callback
```kotlin
val insets = ViewCompat.getRootWindowInsets(binding.root)!!
val isKeyboardVisible = insets.isVisible(Type.ime())
val keyboardHeight = insets.getInsets(Type.ime()).bottom
```
But this function is of no use for API<23
![image](https://user-images.githubusercontent.com/51705072/116391375-b4d08280-a83c-11eb-8e2b-0d713c9e0810.png)

As per the android docs,
> Some caveats about the ime() type, it works very reliably on API 23+ when your Activity is using the adjustResize window soft input mode. If you’re instead using the adjustPan mode, it should work reliably back to API 14.

It says that the API **should** work back to API 14 but I couldn't find the right functions which work below 21.
@rt4914 @BenHenning Any ideas on what functions to use?
<br><br>
@BenHenning Besides this, I had a small doubt. 
- If we are able to make this WindowInsets library work for lower APIs, or 
- if we come up with some UI changes so that it doesn't require checking keyboard's visibility or height.

that would probably be the best solution (second obviously being better).
I just wanted to ask what was actually wrong with the original solution. You said that,
>This seems like a really hacky way to check if the keyboard is open. It's making a lot of assumptions on how large the keyboard is (which is a customizable and OEM-specific property).

I couldn't understand what you meant by **a lot of assumptions**. The only assumption made was for the **minimum** height of a softKeyboard (plus that of system bars). The solution didn't depend on the actual height of keyboard. The exact height is definitely **a customizable and OEM-specific property** but I don't think there is any keyboard which occupies less than 150dp of height. Also, if something is obstructing more than 150dp of device screen, what else can it be other than the keyboard?

Reply:
Is it possible to add "extra bottom padding" to the lesson when keyboard is active? When the keyboard is not active, then the "bottom padding" resets to default. Ideally we would animate the transition so the change in the lesson content's position isn't so jarring.

Reply:
@mschanteltc I don't think so--the problem is that we cannot reliably know if the keyboard is open. I think that Android was designed, in part, to support UIs that don't depend on keyboard-specific behavior.

@ArpitShukIa one issue is that the previous solution was using a 200dp threshold, and even within the stackoverflow discussion linked above there were multiple comments that had subtly different solutions, and others where people mentioned the solution doesn't work on certain devices with different smallest-width or density configurations. That, plus the fact that AndroidX can't properly support this functionality in older API versions with adjustResize I think are fairly strong indicators that trying to detect the keyboard status in this way is fundamentally fragile and likely to result in device-specific bugs (which we really want to try to avoid). Given how large the device ecosystem is for Android (10k's of devices) it's really hard for us to confidently design a size-based solution without the system aiding us, and it seems like the Android team intentionally designed the system to not do that. Does this help clarify?

Regarding with AndroidX API to use, I'm not sure. I found that solution after a search, and you may need to dig more into it to figure out the correct approach for our target API version. Either way, I think it requires us to use adjustPan so we definitely need to decide on whether we should consider a UI redesign or to use adjustPan (or something in-between).

Reply:
Is it possible to shift the latest feedback to the top of the page? Regardless of whether or not the keyboard is active, shifting the latest feedback to the top could allow the user to access the input container, though there still might be cases where the feedback and keyboard are large enough to hide it from view.

Reply:
We should be able to auto-scroll to the feedback, though we may need extra padding to ensure that the content is never underneath the keyboard.

@mschanteltc is your suggestion meant to be for adjustResize or adjustPan for the keyboard?